src/passes/GlobalStructInference.cpp

```diff
@@ -123,6 +123,16 @@ struct GlobalStructInference : public Pass {
 
       auto type = global->init->type.getHeapType();
 
+      // The global's declared type must match the init's type. If not, say if
+      // we had a global declared as type |any| but that contains (ref $A), then
+      // that is not something we can optimize, as ref.eq on a global.get of
+      // that global will not validate. (This should not be a problem after
+      // GlobalSubtyping runs, which will specialize the type of the global.)
```

Comment:
Do we need to worry about the case where `GlobalSubtyping` might refine the global to be some LUB, but it is still refined enough that we could optimize it here?

Reply:
In practice no, since GSI works in a simple way: it connects globals to struct types by seeing `global g = new Type()`, so a 1:1 correspondence. In theory though one day if it's possible to do a `select` in a global init then we might want to support more.